applications/cryptex.md

```diff
@@ -0,0 +1,297 @@
+
+**Genesis/Setup**
+
+  Each validator generates a private key and public key for the underlying signature scheme of the blockchain. Theoretically this could be implemented on any scheme, but we use BLS12-381. Each $A_i \in A$ generates keys $(sk_i, pk_i)$, storing the secret key $sk_i$ securely (in their [keystore](https://paritytech.github.io/substrate/master/sp_keystore/trait.Keystore.html)), with the public keys used to define the initial validators. This is just standard setup.
+   
+  On genesis, each authority must have a known secret available. To make this as simple as possible, we will simply hardcode the genesis secrets within the genesis block. This is the only time the secrets will be encoded in plaintext. 
+
+
+  Before a new session starts and after the initial session, a semi trusted node, the *coordinator*, is responsible for facilitating the threshold secret sharing. It does this by generating a secret polynomial and secret shares, encrypting them, and distributing them to upcoming session validators prior to the session start. To make things even simpler, the coordinator will also derive public keys for each of the authorities based on the secret shares, and publish the public keys along with the encrypted secrets. 
+  
+  The coordinator also chooses system parameters for the IBC, a generator $P \in \mathbb{G}_1$, then calculates $P_{pub} = sP$, where $s$ is the secret created the the TSS scheme. Then publish $(P, P_{pub}, (k_1, Q_1), ..., (k_n, Q_n))$ on-chain, where $k_1, ..., k_n$ and $Q_i$ are the public keys. are the encrypted secrets. This will happen at the end of a session during the `new-session` function.
+
+**Keygen and Identity**
+  Each slot in any given epoch has a unique role associated with it, which is calculated from the slot schedule. For any given address, epoch, and slot number, we calculate a unique role by hashing the address, epoch, and slot number. Later on when encrypting, we will use this value to verify signatures. That is, the public key is $\hat{Q}_i = H(ID_i = (A_i || e_k || sl_r))$.
+
+**Block Sealing**
+  The winner of a slot $s$ calculates a secret key corresponding to $Q_i$ and uses it to sign the block. The winner calculates a signature as $S = s_i \hat{Q}_i + r H(M)$ and $T = r P$, where $M$ is the block hash, $s_i$ is the secret distributed by the coordinator, and $r$ is randomly selected from $\mathbb{Z}_p$. We implement this within the Aura client code.
+
+**Validation**
+  When other nodes import the block, they validate it by  calculating the ID for the slot and calculating the public key $Q_i$ and verifying the signature. We verify the signature by checking the pairings: $e(S, P) = e(Q_i, P_{pub}) + e(H(M), T)$. This validation logic will be executed each time a new block queued for import is validated.
```

Comment:
It's more efficient to verify this without pairings, using a DLEQ proof, like I did above.

`S_slot = s_i Q_i` is the slot secret key, revealed when making the block, although using `i` for both subscripts here gets confusing. sine the `s_i` are really fixed for each collator, aka `s_collator`.

Signer:
```
r <- /dev/urandom
R_1 = r G_1
R_2 = r G_2
pok = c s_i + r 
c = H(...)
```
Verifier:
```
pok G_1 = c pk_i + R_1
pok Q_i = c S_slot + R_2
```
This proves `pk_i = s G_1` and `S_slot = s Q_i` for the same `s`.  These G_2 scalar mults are not much faster than the miller loop, the they avoid the even slower final exponentiation. 

I can show you later how to do this easily using the dleq_vrf crate.

Reply:
That makes a lot of sense, I updated the proposal to reflect that (both sealing and verification). I still have to update the etf_aura.pdf overview to include that; I'll commit that soon.